Add comments to persons

alexshook opened the pull request:
For #272. The original description doesn't say anything about creating comments, only showing them, so I left out create for now but am happy to do it if you give me a few requirements. I added a table to the people show view to display comments. Should this be added anywhere else?

alexshook committed:
Add test for Comment assocation on Person

- Makes sure the Commentable concern is working.

alexshook committed:
Add scope to order Comments from newest to oldest

alexshook committed:
Add Comment validations for person_id, description

Also updates comment factory.

alexshook committed:
Add partial to view comments in a table

Also adds the new partial to the people#show view. I had a hard time
testing that view and ended up not adding tests.

alexshook committed:
Use I18n for the no comments message

alexshook committed:
Order comments by most recent first on the show view

Also eager load the comments association in the controller.

alexshook commented on spec/models/comment_spec.rb:
```diff
@@ -0,0 +1,17 @@
+
+RSpec.describe Comment do
+  describe 'validations' do
+    it { is_expected.to validate_presence_of :person_id }
```
Added some validations before I realized I didn't know what to do about `create`, can remove them if you'd rather have `create` be part of a different story.

alexshook commented on app/controllers/people_controller.rb:
```diff
@@ -73,7 +73,7 @@ def index
 
   def show
-    @person = Person.find(params[:id])
+    @person = Person.includes(:comments).find(params[:id])
```
I waffled a bit between what I did here and going for `@comments = @person.comments`, similar to `@phones` on line 79. Let me know if you have a preference.

kgf replied:
This has the advantage of fewer sql queries.

kgf commented on app/models/comment.rb:
```diff
@@ -1,3 +1,7 @@
 class Comment < ActiveRecord::Base
   belongs_to :commentable, :polymorphic => true
+
+  scope :recent, ->{ order('created_at DESC') }
```
This scope confuses me. It is named `recent`, but it does not limit the comments based on time. Perhaps name it `by_creation_date` or something like that.

kgf replied:
Thinking about it more, you could even make this a default scope, because we would rarely want the unsorted.

alexshook replied:
👍 sounds good to mee

kgf commented:
Sorry I missed your question earlier. Yes, it would be nice to be able to enter them too ! 
Are you doing OK other than that ?

kgf commented on app/models/comment.rb:
```diff
@@ -1,3 +1,7 @@
+
+  scope :recent, ->{ order('created_at DESC') }
+
+  validates_presence_of :person_id, :description
```
You might not have noticed, but this is a polymorphic class. See line 2 of this file. I don't think we can validate on person_id.

alexshook replied:
I believe `person_id` will validate (we should always have the concept of who posted the comment?) since that's stored in the comments table itself. At least I can't think of any reason why it wouldn't, but also haven't touched a polymorphic association in a while. 

I did think that `person_id` might better be called `poster_id`. That might be confusing for when the commentable type and id are a Person though, so... idk. I'll remove the validation if it ends up being a 💣 while implementing create comments.

alexshook committed:
Set default scope for comments to order newest to oldest

https://github.com/ReadyResponder/ReadyResponder/pull/565/files/2a99dc5ffb323b4e6a99fb9104a6462afc0118ef#r144720948

alexshook committed:
Remove scope since it's now the default scope

alexshook commented:
Yes I'm totally good with creating comments as well. Could you give me some
guidance on what you're looking for? Something like where to put a text
field for comments, if you'd like any specific labels for the
fields/placeholder text, etc.? I mostly make APIs so I'm not the best at
deciding where front end elements should go.

kgf commented:
Are you finding this work enjoyable? I could create an issue closer to API work for you.

We can almost merged this here, although it won't quite work. The validation on person id won't work,  as comments are polymorphic.

kgf committed:
Merge branch 'development' into add_comments_to_persons

alexshook committed:
Remove validation for person_id

alexshook commented:
Yeah this is great, always good to do slightly different type of work...
and also to not completely forget how to make a form. I'll remove the
person_id validation now, although I would swear that it would work.


*Alex Shook610.703.8914 | alex.e.shook@gmail.com <alex.e.shook@gmail.com>*
*LinkedIn <http://www.linkedin.com/in/alexshook> | @alexshook
<https://twitter.com/alexshook> | Portfolio <http://alexshook.com/>*

On Wed, Oct 18, 2017 at 10:51 PM, Kevin Fallon <notifications@github.com>
wrote:

> Are you finding this work enjoyable? I could create an issue closer to API
> work for you.
>
> We can almost merged this here, although it won't quite work. The
> validation on person id won't work, as comments are polymorphic.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ReadyResponder/ReadyResponder/pull/565#issuecomment-337784951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFYaS5lhMPTsGbUhiF1b0i2yuvh4iXYAks5strlEgaJpZM4P5lqc>
> .
>

alexshook committed:
Merge branch 'add_comments_to_persons' of github.com:alexshook/ReadyResponder into add_comments_to_persons

alexshook committed:
Also remove the test for the person_id validation

kgf commented:
The person validation might work for this PR, but we're going to add comments to items and events as well. 
Thanks for your help.
Kevin

kgf committed:
Merge branch 'development' into add_comments_to_persons

kgf commented:
Sadly, the tests did not pass...

alexshook commented:
It looks like the certs tests failed, can you rerun the build?

kgf commented:
@alexshook build restarted

kgf committed:
Merge branch 'development' into add_comments_to_persons

kgf commented:
@alexshook thanks for your help. 
This partially addresses #272 .
